stm32h7: add usb hs ulpi support

XenuIsWatching opened the pull request:
Add support for the STM32H7 USB OTG HS and support for the ULPI PHY. Add the ulpi definition example with the STM32H747-Disco Board.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
There is the issue of the pinctrl dtsi slew rate for the ulpi pins are not set right. These are left at the default and these need to be at least "fast-speed"

XenuIsWatching commented:
I have verified this with a custom board... but I don't have a have STM32H747-Disco in front of me yet. I can get one later this week which I can run the test on (unless if someone else is anxious and can test sooner for me :) )

loicpoulain commented:
You should probably split that in several commits.

jfischer-no commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -39,6 +39,19 @@
 
+		usbotg_hs: usb@40040000 {
+			compatible = "st,stm32-otghs";
+			reg = <0x40040000 0x40000>;
```
Is it an other controller on the SoC or why it has a different `reg = <0x40040000 0x40000>;` ?

jfischer-no commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -39,6 +39,19 @@
+			interrupt-names = "ep1_out", "ep1_in", "wkup", "otghs";
+			num-bidir-endpoints = <9>;
+			ram-size = <4096>;
+			maximum-speed = "full-speed";
```
Should it not be `maximum-speed = "high-speed";`? IIRC it does not work in any case without USB_DC_HAS_HS_SUPPORT selected.

XenuIsWatching replied:
it is a high supporting controller.... but it can support 2 PHYs, one is the internal FS PHY which is already within there.... and the other is the HS external ULPI PHY which is to be configured with in another dts.

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -69,6 +69,14 @@ static const struct pinctrl_dev_config *usb_pcfg =
+
+#if USB_OTG_HS_ULPI_PHY
+static const struct gpio_dt_spec ulpi_reset =
+	GPIO_DT_SPEC_GET_OR(DT_PHANDLE(DT_INST(0, st_stm32_otghs), phys), reset_gpios, { 0 });
```
```c
        GPIO_DT_SPEC_GET_OR(DT_PHANDLE(DT_INST(0, st_stm32_otghs), phys), \                                                                  
                            reset_gpios, { 0 });
```

XenuIsWatching replied:
When that is done checkpatch gives an error
```
LINE_CONTINUATIONS: Avoid unnecessary line continuations
File:drivers/usb/device/usb_dc_stm32.c
Line:77
```

jfischer-no reviewed (changes requested):
>You should probably split that in several commits.

Agree. Please split it in two commits, first with changes in bindings and drivers, and second dts and boards.

erwango commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.dts:
```diff
@@ -165,6 +170,26 @@
 
+zephyr_udc0: &usbotg_hs {
+	pinctrl-0 = <&usb_otg_hs_ulpi_ck_pa5
+				&usb_otg_hs_ulpi_d0_pa3
+				&usb_otg_hs_ulpi_d1_pb0
+				&usb_otg_hs_ulpi_d2_pb1
+				&usb_otg_hs_ulpi_d3_pb10
+				&usb_otg_hs_ulpi_d4_pb11
+				&usb_otg_hs_ulpi_d5_pb12
+				&usb_otg_hs_ulpi_d6_pb13
+				&usb_otg_hs_ulpi_d7_pb5
+				&usb_otg_hs_ulpi_stp_pc0
+				&usb_otg_hs_ulpi_dir_pi11
+				&usb_otg_hs_ulpi_nxt_ph4>;
```
If these pins should be set in high-speed, it would make sense to do it directly in hal_stm32 module (modify ulpi pins generation to be high speed by default)

XenuIsWatching replied:
I already have a PR for this within the hal_stm32. I just need to update the module hash to point to the updated commit once that is merged in
https://github.com/zephyrproject-rtos/hal_stm32/pull/155

erwango commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.dts:
```diff
@@ -165,6 +170,26 @@
+				&usb_otg_hs_ulpi_nxt_ph4>;
+	pinctrl-names = "default";
+	maximum-speed = "high-speed";
+	clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x06000000>;
```
A comment stating you enable ULPI PHY would be nice as this is not obvious.

erwango reviewed (changes requested):
Looks globally fine aside from the comments already given which should be taken into account.

Current driver implementation is not great regarding clocks and should be reviewed to make things clearer but this should be done in another PR.

Note that I tried to test on stm32h747_disco_m7 (modifying pin slew rate), no success.

XenuIsWatching commented:
> Looks globally fine aside from the comments already given which should be taken into account.
> 
> Current driver implementation is not great regarding clocks and should be reviewed to make things clearer but this should be done in another PR.
> 
> Note that I tried to test on stm32h747_disco_m7 (modifying pin slew rate), no success.

I wonder if that could be because `CONFIG_USB_DC_HAS_HS_SUPPORT` is not set as @jfischer-no mentioned. I have that enabled in the custom board that I have which is working...I should be able to get ahold of that disco board on Wednesday to test it out on it.

desowin commented:
> > Note that I tried to test on stm32h747_disco_m7 (modifying pin slew rate), no success.
> 
> I wonder if that could be because `CONFIG_USB_DC_HAS_HS_SUPPORT` is not set as @jfischer-no mentioned. I have that enabled in the custom board that I have which is working...I should be able to get ahold of that disco board on Wednesday to test it out on it. I may have to update the KConfig to imply that if that otghs is enabled

Depending on what "no success" means. The `CONFIG_USB_DC_HAS_HS_SUPPORT` only modifies the bulk endpoint max packet size (from 64 to 512, however the old stack does not take the current speed into account). Therefore, if "no success" means "device enumerated but does not work properly" then that can be due to missing `CONFIG_USB_DC_HAS_HS_SUPPORT` but if it means "device doesn't even send the device descriptor to host" then it's different problem.

XenuIsWatching commented:
> CONFIG_USB_DC_HAS_HS_SUPPORT

I got a hold of my STM32H747-Disco.... It works working for me with the code as is and shows up in dmesg
```
[156084.598332] usb 3-6.1: new high-speed USB device number 26 using xhci_hcd
[156084.675055] usb 3-6.1: config 1 interface 1 altsetting 0 bulk endpoint 0x82 has invalid maxpacket 64
[156084.675077] usb 3-6.1: config 1 interface 1 altsetting 0 bulk endpoint 0x1 has invalid maxpacket 64
[156084.676430] usb 3-6.1: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.02
[156084.676444] usb 3-6.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[156084.676450] usb 3-6.1: Product: Zephyr CDC ACM sample
[156084.676456] usb 3-6.1: Manufacturer: ZEPHYR
[156084.676461] usb 3-6.1: SerialNumber: 3438510C00290033
[156084.682997] cdc_acm 3-6.1:1.0: ttyACM0: USB ACM device
```

Although there is the error about the maxpacket which is due to the CONFIG_USB_DC_HAS_HS_SUPPORT missing.
This was with the `samples/subsys/usb/cdc_acm` sample

After adding that Kconfig to be auto-selected if the speed is HS.
```
[156760.938437] usb 3-6.1: new high-speed USB device number 31 using xhci_hcd
[156761.015702] usb 3-6.1: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.02
[156761.015727] usb 3-6.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[156761.015737] usb 3-6.1: Product: Zephyr CDC ACM sample
[156761.015744] usb 3-6.1: Manufacturer: ZEPHYR
[156761.015750] usb 3-6.1: SerialNumber: 3438510C00290033
[156761.019991] cdc_acm 3-6.1:1.0: ttyACM0: USB ACM device
```

XenuIsWatching committed:
drivers: usb: stm32: add usb hs ulpi support

Add support for the STM32H7 USB OTG HS and support for the ULPI PHY.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
boards: stm32h747: add usb otg hs and ulpi phy definition

Add usb otg hs definition for the stm32h747 disco board

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

erwango commented:
> There is the issue of the pinctrl dtsi slew rate for the ulpi pins are not set right. These are left at the default and these need to be at least "high-speed"

Can you add a commit that updates west.yaml and points to commit e1c54e99a40da705c156ae3464a14720e7d668c8 ?

XenuIsWatching committed:
manifest: update hal_stm32

Update hal_stm32 to include the change to set the ulpi
pins to "high-speed" slew rate

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

erwango reviewed (approved):
LGTM.
Tested ok on H7 board
